lxd/devlxd.go

```diff
@@ -560,3 +561,161 @@ func findContainerForPid(pid int32, s *state.State) (instance.Container, error)
 
 	return nil, errPIDNotInContainer
 }
+
+// userKeys is a struct that keeps the keys to be injected to allow SSHing as a certain user.
+type userKeys struct {
+	importIDs  []string
+	publicKeys []string
+}
+
+// mergeSSHKeyCloudConfig merges any existing SSH keys defined in an instance config into a provided
+// cloud-config YAML string.
+// In the case where we were not able to parse the cloud config, return the original, unchanged config and
+// the error.
+func mergeSSHKeyCloudConfig(instanceConfig map[string]string, cloudConfig string) (string, error) {
+	// Use a pointer to userKeys so we can append to its fields.
+	users := make(map[string]*userKeys)
+
+	for key, value := range instanceConfig {
+		if strings.HasPrefix(key, "cloud-init.ssh-keys.") {
+			user, sshKey, found := strings.Cut(value, ":")
+
+			// If the "cloud-init.ssh-keys." is badly formatted, skip it.
+			if !found {
+				continue
+			}
+
+			// Create an empty userKeys if the user is not configured.
+			_, ok := users[user]
+			if !ok {
+				users[user] = &userKeys{}
+			}
+
+			// Check if ssh key is an import ID with with a format "keyServer:UserName".
+			if strings.Contains(sshKey, ":") && !strings.Contains(sshKey, " ") {
+				users[user].importIDs = append(users[user].importIDs, sshKey)
+				continue
+			}
+
+			users[user].publicKeys = append(users[user].publicKeys, sshKey)
+		}
+	}
+
+	// If no additional keys are defined, return the original config passed in.
```

Comment:
"additional" seems misleading here, additional to what? I would suggest we just say "no keys".